From 401e3145577b0c4811dd01ddf848b496bea51ce6 Mon Sep 17 00:00:00 2001
From: "rosen.diankov@gmail.com" <rosen.diankov@mujin.co.jp>
Date: Sat, 7 Oct 2017 12:19:41 +0900
Subject: [PATCH] remove usage of 'const dReal' for gcc7 compat

---
 plugins/rplanners/rplanners.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/plugins/rplanners/rplanners.h b/plugins/rplanners/rplanners.h
index f62582842..77e8b474e 100644
--- a/plugins/rplanners/rplanners.h
+++ b/plugins/rplanners/rplanners.h
@@ -214,7 +214,7 @@ class SpatialTree : public SpatialTreeBase
 
     inline dReal _ComputeDistance(const dReal* config0, const dReal* config1) const
     {
-        return _distmetricfn(VectorWrapper<const dReal>(config0, config0+_dof), VectorWrapper<dReal>(config1, config1+_dof));
+        return _distmetricfn(VectorWrapper<dReal>(config0, config0+_dof), VectorWrapper<dReal>(config1, config1+_dof));
     }
 
     inline dReal _ComputeDistance(const dReal* config0, const std::vector<dReal>& config1) const
